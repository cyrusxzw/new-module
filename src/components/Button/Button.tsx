import React, { forwardRef } from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import type { ButtonProps } from './Button.types';
import styles from './Button.module.css';

const Button = forwardRef<HTMLButtonElement, ButtonProps>(function ButtonRef(
  {
    aria,
    children,
    className,
    dataTestRef,
    id,
    isAlternate = false,
    isEnabled = true,
    isInline = false,
    onClick,
    onMouseEnter,
    tabIndex,
    title,
    type = 'button',
    theme,
    ...otherProps
  },
  ref,
) {
  const currentTheme = useThemeContext(theme, 'dark');

  if (!children) return null;

  const classSet = cx(
    styles.base,
    { [styles.alternate]: isAlternate },
    { [styles.blockStyle]: !isInline },
    styles[currentTheme],
    { [styles.disabled]: !isEnabled },
    className,
  );

  return (
    <button
      aria-expanded={aria?.expanded ?? null}
      aria-haspopup={aria?.haspopup ?? null}
      aria-hidden={aria?.hidden ?? null}
      aria-label={aria?.label ?? null}
      className={classSet}
      data-test-ref={dataTestRef}
      disabled={!isEnabled}
      id={id}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      ref={ref}
      tabIndex={tabIndex}
      title={title}
      type={type}
      {...otherProps}
    >
      {children}
    </button>
  );
});

export { Button };
