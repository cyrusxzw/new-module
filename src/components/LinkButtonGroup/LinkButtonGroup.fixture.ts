const LinkButtonGroupFixture = {
  link: {
    text: 'Lorem ipsum',
    url: '/',
    style: 'Internal No Icon Button Link',
  },
  secondaryLink: {
    text: 'Nunc dignissim quis',
    url: '/',
    style: 'External No Icon Button Link',
  },
};

export { LinkButtonGroupFixture };
