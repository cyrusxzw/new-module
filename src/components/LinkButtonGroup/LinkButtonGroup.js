import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import styles from './LinkButtonGroup.module.css';

const isReactFragment = (component) => {
  if (component.type) return component.type === Fragment;

  return component === Fragment;
};

const LinkButtonGroup = ({
  children,
  className,
  hasFitContent,
  isFlush,
  isFullWidth,
  textAlign,
  theme,
}) => {
  const currentTheme = useThemeContext(theme, 'dark');

  if (!children || typeof children === 'string' || children instanceof String) {
    return null;
  }

  const classSet = cx(
    styles.base,
    { [styles.fitContent]: hasFitContent },
    { [styles.flush]: isFlush },
    className,
  );

  const childrenClassSet = cx(
    styles.link,
    { [styles.flushLink]: isFlush },
    { [styles.fitContent]: hasFitContent },
    { [styles.fullWidth]: isFullWidth },
    styles[textAlign],
  );

  const childComponents = isReactFragment(children)
    ? children.props.children
    : children;

  if (childComponents === undefined) return null;

  return (
    <div className={classSet} data-testid="data-testid-LinkButtonGroup">
      <>
        {React.Children.map(childComponents, (child) => {
          if (child === null) return null;

          return React.cloneElement(child, {
            className: `${child.props?.className || ''} ${childrenClassSet}`,
            textAlign,
            theme: currentTheme,
          });
        })}
      </>
    </div>
  );
};

LinkButtonGroup.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.arrayOf(PropTypes.element),
  ]),
  className: PropTypes.string,
  hasFitContent: PropTypes.bool,
  isFlush: PropTypes.bool,
  isFullWidth: PropTypes.bool,
  textAlign: PropTypes.oneOf(['center', 'left', 'right']),
  theme: PropTypes.oneOf(['dark', 'light']),
};

LinkButtonGroup.defaultProps = {
  children: undefined,
  className: undefined,
  hasFitContent: false,
  isFlush: true,
  isFullWidth: false,
  textAlign: 'center',
  theme: undefined,
};

export { LinkButtonGroup };
