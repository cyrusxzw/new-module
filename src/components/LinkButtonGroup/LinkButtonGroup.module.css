@import 'styles/index';

.base {
  display: flex;
  max-width: 100%;
  flex-wrap: wrap;
  margin-right: -2px;
  margin-left: -2px;

  &,
  & * {
    box-sizing: border-box;
  }
}

.flush {
  width: calc(100% - 1px);
  margin-right: 0;
  margin-left: 0;
}

.link {
  display: flex;
  width: auto;
  min-width: 210px;
  max-width: 300px;
  flex: 1 1 auto;
  align-items: center;
  margin-right: 2px;
  margin-bottom: 10px;
  margin-left: 2px;
  font-size: rem(14px);
  line-height: 1.4;

  @media (--viewport-xs-sm-only) {
    width: auto;
    max-width: 100%;
  }

  @media (--viewport-lg) {
    margin-bottom: 20px;
  }
}

.flushLink {
  margin-right: -0.5px;
  margin-left: -0.5px;
}

.fitContent {
  width: fit-content;
}

.fullWidth.fullWidth {
  max-width: none;
}
