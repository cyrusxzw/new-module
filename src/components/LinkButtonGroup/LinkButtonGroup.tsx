import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import type { LinkButtonGroupProps } from './LinkButtonGroup.types';
import styles from './LinkButtonGroup.module.css';

const LinkButtonGroup = ({
  children,
  className,
  hasFitContent,
  isFlush = true,
  isFullWidth = false,
  textAlign = 'center',
  theme,
}: LinkButtonGroupProps): React.ReactElement => {
  const currentTheme = useThemeContext(theme, 'dark');

  const classSet = cx(
    styles.base,
    { [styles.fitContent]: hasFitContent },
    { [styles.flush]: isFlush },
    className,
  );

  const childrenClassSet = cx(
    styles.link,
    { [styles.flushLink]: isFlush },
    { [styles.fitContent]: hasFitContent },
    { [styles.fullWidth]: isFullWidth },
    styles[textAlign],
  );

  return (
    <div className={classSet} data-testid="data-testid-LinkButtonGroup">
      <>
        {React.Children.map(children, (child) => {
          if (child === null) return null;

          return React.cloneElement(child, {
            className: `${child.props?.className || ''} ${childrenClassSet}`,
            textAlign,
            theme: currentTheme,
          });
        })}
      </>
    </div>
  );
};

export { LinkButtonGroup };
