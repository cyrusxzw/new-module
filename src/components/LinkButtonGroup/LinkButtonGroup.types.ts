import type { ReactElement } from 'react';
import type { Themes } from '~/types';

type LinkButtonGroupProps = {
  children?:
    | JSX.Element
    | JSX.Element[] /** @todo figure out how to efficiently allow only Button an Hyperlink */;
  className?: string;
  hasFitContent?: boolean;
  isFlush?: boolean;
  isFullWidth?: boolean;
  textAlign?: 'center' | 'left' | 'right';
  theme?: Themes;
};

type LinkButtonGroupType = (props: LinkButtonGroupProps) => ReactElement | null;

export type { LinkButtonGroupProps, LinkButtonGroupType };
