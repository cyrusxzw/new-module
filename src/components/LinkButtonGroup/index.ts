import { LinkButtonGroup } from './LinkButtonGroup';

export { LinkButtonGroup };
