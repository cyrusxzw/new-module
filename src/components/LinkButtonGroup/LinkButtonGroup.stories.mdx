import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import * as knobs from '@storybook/addon-knobs';
import { StorybookGridOverlay } from '~.storybook/components';
import { Hyperlink } from '~/components/Hyperlink';
import { Button } from '~/components/Button';
import { LinkButtonGroup } from './LinkButtonGroup';
import { LinkButtonGroupFixture } from './LinkButtonGroup.fixture';

<Meta title="Components/LinkButtonGroup" component={LinkButtonGroup} />

# LinkButtonGroup

<Canvas>
  <Story name="Base component">
    {(() => {
      const theme = knobs.select(
        'theme',
        ['dark', 'light'],
        'dark',
        'Presentation',
      );
      const implementation = {
        gridOverlay: {
          isActive: knobs.boolean('Show grid overlay', false, 'Implementation'),
          hasInvertedColours: knobs.boolean(
            'Invert grid overlay colours',
            false,
            'Implementation',
          ),
        },
      };
      return (
        <>
          <LinkButtonGroup
            hasFitContent={knobs.boolean('hasFitContent', false, 'Presentation')}
            theme={knobs.select('theme', ['dark', 'light'], 'dark', 'Presentation')}
            textAlign={knobs.select(
              'textAlign',
              ['center', 'left', 'right'],
              'center',
              'Presentation',
            )}
          >
            <Hyperlink
              url={knobs.text('url', LinkButtonGroupFixture.link.url, 'First link')}
              style={knobs.select(
                'style',
                [
                  'External Button Link',
                  'Internal Button Link',
                  'External No Icon Button Link',
                  'Internal No Icon Button Link',
                ],
                LinkButtonGroupFixture.link.style,
                'First link',
              )}
            >
              {knobs.text(
                'children',
                LinkButtonGroupFixture.link.text,
                'First link',
              )}
            </Hyperlink>
            <Hyperlink
              url={knobs.text('url', LinkButtonGroupFixture.secondaryLink.url, 'First link')}
              style={knobs.select(
                'style',
                [
                  'External Button Link',
                  'Internal Button Link',
                  'External No Icon Button Link',
                  'Internal No Icon Button Link',
                ],
                LinkButtonGroupFixture.secondaryLink.style,
                'First link',
              )}
            >
              {knobs.text(
                'children',
                LinkButtonGroupFixture.secondaryLink.text,
                'First link',
              )}
            </Hyperlink>
          </LinkButtonGroup>
          <StorybookGridOverlay
            isActive={implementation.gridOverlay.isActive}
            hasInvertedColours={implementation.gridOverlay.hasInvertedColours}
          />
        </>
      );
    })()}
  </Story>
</Canvas>

## API

<ArgsTable of={LinkButtonGroup} />
