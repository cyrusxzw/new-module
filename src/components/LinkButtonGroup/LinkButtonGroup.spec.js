import React from 'react';
import { render, screen } from '@testing-library/react';
import { LinkButtonGroup } from './LinkButtonGroup';
import { Button } from '~/components/Button';

describe('<LinkButtonGroup />', () => {
  it('renders children correctly', () => {
    const { container } = render(
      <LinkButtonGroup theme="light">
        <Button onClick={() => undefined} title="test button">
          Test
        </Button>
      </LinkButtonGroup>,
    );

    const child = screen.getByTestId(/data-testid-LinkButtonGroup/i);

    expect(container).toMatchSnapshot();
    expect(child).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /test button/i })).toBeTruthy();
  });

  it('should return `null` if no children were passed', () => {
    render(<LinkButtonGroup />);

    const child = screen.queryByTestId(/data-testid-LinkButtonGroup/i);

    expect(child).not.toBeInTheDocument();
  });

  it('should return `null` if only a fragment was passed', () => {
    render(
      <LinkButtonGroup>
        <></>
      </LinkButtonGroup>,
    );

    const child = screen.queryByTestId(/data-testid-LinkButtonGroup/i);

    expect(child).not.toBeInTheDocument();
  });

  it('should render correctly if first direct descendant is a fragment', () => {
    render(
      <LinkButtonGroup textAlign="right" theme="light">
        <>
          <Button onClick={() => undefined} title="test button">
            Test
          </Button>
        </>
      </LinkButtonGroup>,
    );

    const child = screen.getByTestId(/data-testid-LinkButtonGroup/i);

    expect(child).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /test button/i })).toBeTruthy();
  });
});
