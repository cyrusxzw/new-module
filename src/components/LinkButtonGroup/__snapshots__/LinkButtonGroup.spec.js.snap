// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LinkButtonGroup /> renders children correctly 1`] = `
<div>
  <div
    class="base flush"
    data-testid="data-testid-LinkButtonGroup"
  >
    <button
      class="base blockStyle light  link flushLink center"
      title="test button"
      type="button"
    >
      Test
    </button>
  </div>
</div>
`;
